### PR TITLE
Optimized HNSW heuristic

### PR DESCRIPTION
**Pull request by IvanPleshkov**
### All Submissions:

* [ ] Have you followed the guidelines in our Contributing document?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [ ] Does your submission pass tests?
2. [ ] Have you formatted your code locally using `cargo fmt` command prior to submission?
3. [ ] Have you checked your code using `cargo clippy` command?

### Changes to Core Features:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: lib/segment/src/index/hnsw_index/graph_layers_builder.rs [hunk 1]
```diff
@@ -234,23 +250,66 @@ impl GraphLayersBuilder {
...
-            links.insert(id_to_insert, new_point_id);
+        if links.links.len() < level_m {
+            links.links.insert(id_to_insert, new_point_id);
+        } else if id_to_insert != links.links.len() {
+            links.links.pop();
+            links.links.insert(id_to_insert, new_point_id);
         }
     }
 
+    fn select_one_candidate_with_heuristic_from_sorted<F>(
```
Comment (coszio) [inline lib/segment/src/index/hnsw_index/graph_layers_builder.rs:270]: What's the main difference with respect to `select_candidate_with_heuristic_from_sorted`? I may not be in the loop here but the naming stands out a little ambiguous.

File: lib/segment/src/index/hnsw_index/graph_layers_builder.rs [hunk 2]
```diff
@@ -836,4 +919,69 @@ mod tests {
         graph_layers_builder.links_map(0, 0, |link| result.push(link));
         assert_eq!(&result, &vec![1, 2, 3, 4, 5, 6]);
     }
+
+    #[test]
+    fn test_optimized_heuristic() {
```
Comment (coszio) [inline lib/segment/src/index/hnsw_index/graph_layers_builder.rs:924]: Are we keeping both the optimized + unoptimized versions available? Though it's nice that we can compare them for making sure they behave the same

**Review by coszio — COMMENTED** [review]
Still trying to wrap my head around this 🧠, left a few overview comments

File: lib/segment/src/index/hnsw_index/graph_layers_builder.rs (hunk 1 above)
Comment (IvanPleshkov) [reply lib/segment/src/index/hnsw_index/graph_layers_builder.rs:270]: I tried to explain main idea of PR in notions
https://www.notion.so/qdrant/Heuristic-optimization-ee4ff1ee35cc4ce2828ef87d09dd300b

File: lib/segment/src/index/hnsw_index/graph_layers_builder.rs (hunk 2 above)
Comment (IvanPleshkov) [reply lib/segment/src/index/hnsw_index/graph_layers_builder.rs:924]: This change is very sensitive and I have to be sure that graph didn't change

**Comment by IvanPleshkov** [comment]
Close this PR because it's simpler to reimplement this idea than fix all conflicts. Also, benchmarks are required

**Pull request closed without merging**